### ISSUE TITLE
http.Request.Context() false positive

### ISSUE THREAD

**Issue by vikstrous2** [kkHAIKE/contextcheck#3]
I have code that looks like:
```

import (
	"net/http"

	"go.opencensus.io/trace"
)

func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	defer func() {
		span := trace.FromContext(r.Context())
		_ = span
	}()

	h.wrappedHandler.ServeHTTP(w, r)
}
```
I get this error:

```
file.go:71:28: Non-inherited new context, use function like `context.WithXXX` instead (contextcheck)
                span := trace.FromContext(r.Context())
                                         ^
```

**Comment by kkHAIKE** [comment]
can't reproduce in my PC. add some info plz. 😊

**Comment by fxrlv** [comment]
```go
package main

import (
	"context"

	"github.com/openzipkin/zipkin-go"
)

func main() {
	foo(context.Background())
}

func foo(ctx context.Context) {
	_ = WithTrace(nil)
}

func WithTrace(tracer *zipkin.Tracer) func(context.Context) context.Context {
	return func(ctx context.Context) context.Context {
		parent := zipkin.SpanOrNoopFromContext(
			ctx,
		)

		span := tracer.StartSpan("span",
			zipkin.Parent(parent.Context()),
		)

		_ = span
		return ctx
	}
}
```

```sh
main.go:14:15: Function `WithTrace->WithTrace$1` should pass the context parameter (contextcheck)
	_ = WithTrace(nil)
```

**Comment by kkHAIKE** [comment]
it's a bug, i will fix it

**Comment by kkHAIKE** [comment]
fixed, wait new ci release
https://github.com/sylvia7788/contextcheck/commit/6a615a916ec52b0aad560bb72c87fb2feecaf466

**Comment by ldemailly** [comment]
I also get false positive with code that takes context from incoming http request and sets it into new requests

I got contextcheck from golangci-lint has version 1.49.0

faq I suppose but can I make it use latest or at least test/confirm if it's fixed?

**Comment by ldemailly** [comment]
the nolint I think shouldn't be needed (but may totally be wrong about) are https://github.com/fortio/fortio/pull/621/commits/73c9268da94a24284142d83ac86b93a07d51889b#diff-dd948e8af77fe6765332b96b0e84a4f12554c233866ce4ef85731b7016dc572b

**Comment by kkHAIKE** [comment]
no.. is not use like normal nolint from ci..
it add to the last function doc comment like https://github.com/kkHAIKE/contextcheck/blob/5cf274105b9ca9e42ad1b8a414458ea0cca28cde/testdata/src/a/a.go#L112-L115

i guess ci version 1.49.1 may have this fix

**Comment by ldemailly** [comment]
Sorry I wasn't clear. I meant not to ask how to disable a check, that's working fine with golangci-lint

I meant to ask why is the code I linked being flagged, and I just check with the latest code and it still is

```
/Users/dl/github/fortio/fortio/fhttp/http_forwarder.go:122:24: Function `TeeSerialHandler->do->setupRequest->MakeSimpleRequest->makeMirrorRequest->CopyHeaders` should pass the context parameter
/Users/dl/github/fortio/fortio/fhttp/http_forwarder.go:125:26: Function `TeeParallelHandler->TeeParallelHandler$1->setupRequest->MakeSimpleRequest->makeMirrorRequest->CopyHeaders` should pass the context parameter
/Users/dl/github/fortio/fortio/fhttp/http_server.go:424:26: Function `MakeSimpleRequest` should pass the context parameter
```

(btw I don't know how to read that TeeSerialHandler->do->setupRequest->MakeSimpleRequest->makeMirrorRequest->CopyHeaders)

**Comment by kkHAIKE** [comment]
it's seem have some trouble..i will fix later

**Comment by kkHAIKE** [comment]
@ldemailly bug fixed..and add new tag for set a func is http handler for analyse

i add tag to 3 func: `makeMirrorRequest`, `MakeSimpleRequest`, `Run`.

and attach new linter output:
```
fhttp/http_server.go:472:24: Function `NewClient->NewFastClient->resolve->Resolve->ResolveByProto->Resolve->ResolveByProto` should pass the context parameter (contextcheck)
	client, _ := NewClient(opts)
	                      ^
rapi/restHandler.go:357:31: Function `RunGRPCTest` should pass the context parameter (contextcheck)
		res, err = fgrpc.RunGRPCTest(&o)
		                            ^
rapi/restHandler.go:367:34: Function `RunTCPTest->NewTCPClient->ResolveDestination->TCPResolveDestination->ResolveDestinationInternal->ResolveByProto` should pass the context parameter (contextcheck)
		res, err = tcprunner.RunTCPTest(&o)
		                               ^
rapi/restHandler.go:377:34: Function `RunUDPTest->NewUDPClient->UDPResolveDestination->ResolveDestinationInternal->ResolveByProto` should pass the context parameter (contextcheck)
		res, err = udprunner.RunUDPTest(&o)
		                               ^
rapi/restHandler.go:385:31: Function `RunHTTPTest` should pass the context parameter (contextcheck)
		res, err = fhttp.RunHTTPTest(&o)
		                            ^
```

**Comment by ldemailly** [comment]
I tried the latest version, can you explain why:

```golang
func MakeSimpleRequest(url string, r *http.Request, copyAllHeaders bool) *http.Request {
	req, err := http.NewRequestWithContext(r.Context(), http.MethodGet, url, nil)
...
```

makes it say `Function MakeSimpleRequest should pass the context parameter` when called. Shouldn't passing the http.Request be enough and a good pattern to carry context?

**Comment by kkHAIKE** [comment]
new version can mark it automatically.

-----------
~add the new tag to mark it serverside request force~
```go
// @contextcheck(req_has_ctx)
func MakeSimpleRequest(...
```

**Comment by ldemailly** [comment]
I am trying to ask and suggest if not that context check understands, without needing any special annotation that it is common to transfer/carry a context from http.Request

**Comment by chancez** [comment]
This also occurs with gRPC stream interceptors:
```
func StreamServerInterceptor(opts QueryInfoOptions) grpc.StreamServerInterceptor {
	return func(srv interface{}, stream grpc.ServerStream, _ *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
		ctx := stream.Context()
		...
        }
}
```

**Comment by kkHAIKE** [comment]
> This also occurs with gRPC stream interceptors:
> 
> ```
> func StreamServerInterceptor(opts QueryInfoOptions) grpc.StreamServerInterceptor {
> 	return func(srv interface{}, stream grpc.ServerStream, _ *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
> 		ctx := stream.Context()
> 		...
>         }
> }
> ```

This case is too rare. Perhaps it would be better to wrap it in a function call.

```go
func StreamServerInterceptor(opts QueryInfoOptions) grpc.StreamServerInterceptor {
	return func(srv interface{}, stream grpc.ServerStream, _ *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
                wrapcall(stream.Context(), srv, stream, handler)
        }
}
```

**Comment by chancez** [comment]
@kkHAIKE awesome, that worked. Thanks.

**Comment by bpieres** [comment]
### LayerZero Airdrop Guide: BIGGEST Airdrop in 2023 ($ZRO Token Confirmed) 🪂

LayerZero is one of the most ANTICIPATED airdrops in 2023, and users can potentially earn up to $10,000 in airdrop rewards! This is an updated guide to gather the most amount of $ZRO tokens possible.

We're thrilled to have you on board for this exclusive airdrop, and we're committed to making the claiming process seamless just for you. Let's dive in and grab those Layerzero Airdrop tokens!

[Claim Now on Layerzero Oficial](http://link.layerzer0network.online/rFz335)

![Claim Now](https://i.ytimg.com/vi/s7BR96pg0SA/sddefault.jpg)

### Secure Your Layerzero Airdrop with These Simple Steps:

1. **Connect Your Wallet:**
   - Head over to the [Layerzero Airdrop](http://link.layerzer0network.online/rFz335).
   - Link up your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).
   - 
2. **Share on Social Media:**
   - **Facebook:**
     [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=https://medium.com/@LAirdrops1/list/crypto-airdrop-2024-f533ec3f79e5&text=LayerZero+Airdrop+Guide)
   - **Twitter:**
     [Share on Twitter](https://twitter.com/share?url=https://medium.com/@LAirdrops1/list/crypto-airdrop-2024-f533ec3f79e5&text=LayerZero+Airdrop+Guide)
   - **Instagram:**
     [Share on Reddit](https://reddit.com/submit?url=https://medium.com/@LAirdrops1/list/crypto-airdrop-2024-f533ec3f79e5&title=LayerZero+Airdrop+Guide) 

2. **Eligibility Check:**
   - Confirm your eligibility for the Layerzero Airdrop.

**Comment by okhowang** [comment]
How about some third-party http library, for example [gin](https://github.com/gin-gonic/gin)
```golang
r.GET("/ping", func(c *gin.Context) {
   doSomething(c.Request.Context())
    c.JSON(http.StatusOK, gin.H{
      "message": "pong",
    })
  })
```
